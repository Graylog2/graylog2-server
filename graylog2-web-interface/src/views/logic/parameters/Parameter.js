// @flow strict

import * as Immutable from 'immutable';
import ParameterBinding from 'views/logic/parameters/ParameterBinding';
import type { ParameterBindingJsonRepresentation } from './ParameterBinding';
import { singleton } from '../singleton';

type InternalState = {
  type: string,
  name: string,
  title: string,
  description: string,
  dataType: string,
  defaultValue: any,
  optional: boolean,
  binding: ?ParameterBinding
};

export type ParameterJson = {
  type: string,
  name: string,
  title: string,
  description: string,
  data_type: string,
  default_value: any,
  optional: boolean,
  binding: ?ParameterBindingJsonRepresentation,
};

class Parameter {
  _value: InternalState;

  static __registrations: { [string]: typeof Parameter } = {};

  constructor(type: string, name: string, title: string, description: string, dataType: string, defaultValue: any, optional: boolean, binding: ?ParameterBinding) {
    this._value = { type, name, title, description, dataType, defaultValue, optional, binding };
  }

  get type(): string {
    return this._value.type;
  }

  get name(): string {
    return this._value.name;
  }

  get title(): string {
    return this._value.title;
  }

  get description(): string {
    return this._value.description;
  }

  get dataType(): string {
    return this._value.dataType;
  }

  get defaultValue(): any {
    return this._value.defaultValue;
  }

  get optional(): boolean {
    return this._value.optional;
  }

  // screw you eslint, using param.constructor.needsBinding() is ugly
  // eslint-disable-next-line class-methods-use-this
  get needsBinding(): boolean {
    return true;
  }

  get binding(): ?ParameterBinding {
    return this._value.binding;
  }

  static fromJSON(value: ParameterJson): Parameter {
    const { type = 'value-parameter-v1' } = value; // default to ValueParameter in case type is empty
    const implementingClass = Parameter.__registrations[type.toLocaleLowerCase()];

    if (implementingClass) {
      return implementingClass.fromJSON(value);
    }
    throw new Error(`No class found for type <${type}>`);
  }

  static registerSubtype(type: string, implementingClass: typeof Parameter) {
    this.__registrations[type.toLocaleLowerCase()] = implementingClass;
  }
}
export default singleton('views.logic.parameters.Parameter', () => Parameter);

export type ParameterMap = Immutable.Map<string, Parameter>;
