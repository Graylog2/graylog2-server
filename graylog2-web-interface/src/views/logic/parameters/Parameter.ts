/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */

import * as Immutable from 'immutable';

import ParameterBinding from 'views/logic/parameters/ParameterBinding';

import type { ParameterBindingJsonRepresentation } from './ParameterBinding';

import { singleton } from '../singleton';

type InternalState = {
  type: string,
  name: string,
  title: string,
  description: string,
  dataType: string,
  defaultValue: any,
  optional: boolean,
  binding: ParameterBinding | undefined | null,
};

export type ParameterJson = {
  type: string,
  name: string,
  title: string,
  description: string,
  data_type: string,
  default_value: any,
  optional: boolean,
  binding: ParameterBindingJsonRepresentation | undefined | null,
};

type ParameterSubClass = {
  fromJSON: (json: ParameterJson) => Parameter;
};

class Parameter {
  protected _value: InternalState;

  static __registrations: { [key: string]: ParameterSubClass } = {};

  constructor(type: string, name: string, title: string, description: string, dataType: string, defaultValue: any, optional: boolean, binding?: ParameterBinding) {
    this._value = { type, name, title, description, dataType, defaultValue, optional, binding };
  }

  get type(): string {
    return this._value.type;
  }

  get name(): string {
    return this._value.name;
  }

  get title(): string {
    return this._value.title;
  }

  get description(): string {
    return this._value.description;
  }

  get dataType(): string {
    return this._value.dataType;
  }

  get defaultValue(): any {
    return this._value.defaultValue;
  }

  get optional(): boolean {
    return this._value.optional;
  }

  // screw you eslint, using param.constructor.needsBinding() is ugly
  // eslint-disable-next-line class-methods-use-this
  get needsBinding(): boolean {
    return true;
  }

  get binding(): ParameterBinding | undefined | null {
    return this._value.binding;
  }

  static fromJSON(value: ParameterJson): Parameter {
    const { type = 'value-parameter-v1' } = value; // default to ValueParameter in case type is empty
    const implementingClass = Parameter.__registrations[type.toLocaleLowerCase()];

    if (implementingClass) {
      return implementingClass.fromJSON(value);
    }

    throw new Error(`No class found for type <${type}>`);
  }

  static registerSubtype(type: string, implementingClass: ParameterSubClass) {
    this.__registrations[type.toLocaleLowerCase()] = implementingClass;
  }
}

const SingletonParameter = singleton('views.logic.parameters.Parameter', () => Parameter);
// eslint-disable-next-line @typescript-eslint/no-redeclare
type SingletonParameter = InstanceType<typeof Parameter>;

export default SingletonParameter;

export type ParameterMap = Immutable.Map<string, Parameter>;
