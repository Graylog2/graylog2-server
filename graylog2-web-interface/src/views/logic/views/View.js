// @flow strict
import * as Immutable from 'immutable';
import ObjectID from 'bson-objectid';

import ViewState from './ViewState';
import Search from '../search/Search';
import type { QueryId } from '../queries/Query';

export type Properties = Immutable.List<any>;

export type PluginMetadata = {
  name: string,
  url: string,
};
export type Requirements = { [string]: PluginMetadata };
export type ViewStateMap = Immutable.Map<QueryId, ViewState>;

export type SearchType = 'SEARCH';
export type DashboardType = 'DASHBOARD';
export type ViewType = SearchType | DashboardType;

type InternalState = {
  id: string,
  type: ViewType,
  title: string,
  summary: string,
  description: string,
  search: Search,
  properties: Properties,
  state: ViewStateMap,
  createdAt: Date,
  owner: string,
  requires: Requirements,
};

export type WidgetMapping = Immutable.Map<string, string>;
export type ViewJson = {
  id: string,
  type: ViewType,
  title: string,
  summary: string,
  description: string,
  search_id: string,
  properties: Properties,
  state: { [QueryId]: ViewState },
  created_at: Date,
  owner: string,
  requires: Requirements,
};

export default class View {
  static Type: { Search: SearchType, Dashboard: DashboardType } = {
    Search: 'SEARCH',
    Dashboard: 'DASHBOARD',
  };

  _value: InternalState;

  constructor(id: string,
    type: ViewType,
    title: string,
    summary: string,
    description: string,
    search: Search,
    properties: Properties,
    state: ViewStateMap,
    createdAt: Date,
    owner: string,
    requires: Requirements) {
    this._value = {
      id,
      type,
      title,
      summary,
      description,
      search,
      properties: Immutable.fromJS(properties),
      state: Immutable.fromJS(state),
      createdAt,
      owner,
      requires,
    };
  }

  static create(): View {
    // eslint-disable-next-line no-use-before-define
    return new Builder().createdAt(new Date()).build();
  }

  get id(): string {
    return this._value.id;
  }

  get type(): ViewType {
    return this._value.type;
  }

  get title(): string {
    return this._value.title;
  }

  get summary(): string {
    return this._value.summary;
  }

  get description(): string {
    return this._value.description;
  }

  get search(): Search {
    return this._value.search;
  }

  get properties(): Properties {
    return this._value.properties;
  }

  get state(): ViewStateMap {
    return this._value.state;
  }

  get createdAt(): Date {
    return this._value.createdAt;
  }

  get widgetMapping(): WidgetMapping {
    return (this.state || Immutable.Map()).valueSeq().map(s => s.widgetMapping).reduce((prev, cur) => Immutable.fromJS(prev).merge(Immutable.fromJS(cur)));
  }

  get owner(): string {
    return this._value.owner;
  }

  get requires(): Requirements {
    return this._value.requires || {};
  }

  // eslint-disable-next-line no-use-before-define
  toBuilder(): Builder {
    const { id, title, summary, description, search, properties, state, createdAt, owner, requires, type } = this._value;
    // eslint-disable-next-line no-use-before-define
    return new Builder(Immutable.Map({
      id,
      title,
      summary,
      description,
      search,
      properties,
      state,
      createdAt,
      owner,
      requires,
      type,
    }));
  }

  toJSON() {
    const { id, type, title, summary, description, search, properties, state, createdAt, owner } = this._value;

    return {
      id,
      type,
      title,
      summary,
      description,
      search_id: search.id,
      properties,
      state,
      created_at: createdAt,
      owner,
    };
  }

  static fromJSON(value: ViewJson): View {
    // eslint-disable-next-line camelcase
    const { id, type, title, summary, description, properties, state, created_at, owner, requires } = value;
    const viewState: ViewStateMap = Immutable.Map(state).map(ViewState.fromJSON);
    return View.create()
      .toBuilder()
      .id(id)
      .type(type)
      .title(title)
      .summary(summary)
      .description(description)
      .properties(properties)
      .state(viewState)
      .createdAt(created_at)
      .owner(owner)
      .requires(requires)
      .build();
  }

  // eslint-disable-next-line no-use-before-define
  static builder(): Builder {
    // eslint-disable-next-line no-use-before-define
    return new Builder();
  }
}

class Builder {
  value: Immutable.Map<string, any>;

  constructor(value: Immutable.Map = Immutable.Map()) {
    this.value = value;
  }

  id(value: string): Builder {
    return new Builder(this.value.set('id', value));
  }

  toNewView(): Builder {
    return new Builder(this.value.set('id', undefined).set('title', undefined));
  }

  type(value: ViewType): Builder {
    return new Builder(this.value.set('type', value));
  }

  newId(): Builder {
    return this.id(ObjectID().toString());
  }

  title(value: string): Builder {
    return new Builder(this.value.set('title', value));
  }

  summary(value: string): Builder {
    return new Builder(this.value.set('summary', value));
  }

  description(value: string): Builder {
    return new Builder(this.value.set('description', value));
  }

  search(value: Search): Builder {
    return new Builder(this.value.set('search', value));
  }

  properties(value: Properties): Builder {
    return new Builder(this.value.set('properties', value));
  }

  state(value: ViewStateMap): Builder {
    return new Builder(this.value.set('state', Immutable.fromJS(value)));
  }

  createdAt(value: Date): Builder {
    return new Builder(this.value.set('createdAt', value));
  }

  owner(value: string): Builder {
    return new Builder(this.value.set('owner', value));
  }

  requires(value: Requirements): Builder {
    return new Builder(this.value.set('requires', value));
  }

  build(): View {
    const { id, type, title, summary, description, search, properties, state, createdAt, owner, requires } = this.value.toObject();
    return new View(id, type, title, summary, description, search, properties, state, createdAt, owner, requires);
  }
}
