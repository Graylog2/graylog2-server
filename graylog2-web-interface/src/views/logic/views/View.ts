/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as Immutable from 'immutable';
import { flatten } from 'lodash';
import ObjectID from 'bson-objectid';

import Widget from 'views/logic/widgets/Widget';
import defaultTitle from 'views/components/defaultTitle';

import ViewState from './ViewState';
import type { WidgetMapping } from './types';
import type { ViewStateJson } from './ViewState';

import Search from '../search/Search';
import type { SearchType as QuerySearchType } from '../queries/SearchType';
import type { QueryId } from '../queries/Query';

export type Properties = Immutable.List<any>;

export type PluginMetadata = {
  name: string,
  url: string,
};
export type Requirements = { [key: string]: PluginMetadata };
export type ViewStateMap = Immutable.Map<QueryId, ViewState>;

export type SearchType = 'SEARCH';
export type DashboardType = 'DASHBOARD';
export type ViewType = SearchType | DashboardType;

type InternalState = {
  id: string,
  type: ViewType,
  title: string,
  summary: string,
  description: string,
  search: Search,
  properties: Properties,
  state: ViewStateMap,
  createdAt: Date,
  owner: string,
  requires: Requirements,
};

export type ViewJson = {
  id: string,
  type: ViewType,
  title: string,
  summary: string,
  description: string,
  search_id: string,
  properties: Properties,
  state: { [key: string]: ViewStateJson },
  created_at: Date,
  owner: string,
  requires: Requirements,
};

export default class View {
  static Type: { Search: SearchType, Dashboard: DashboardType } = {
    Search: 'SEARCH',
    Dashboard: 'DASHBOARD',
  };

  _value: InternalState;

  constructor(id: string,
    type: ViewType,
    title: string,
    summary: string,
    description: string,
    search: Search,
    properties: Properties,
    state: ViewStateMap,
    createdAt: Date,
    owner: string,
    requires: Requirements) {
    this._value = {
      id,
      type,
      title,
      summary,
      description,
      search,
      properties: Immutable.fromJS(properties),
      state: Immutable.fromJS(state),
      createdAt,
      owner,
      requires,
    };
  }

  static create(): View {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder().createdAt(new Date()).build();
  }

  get id(): string {
    return this._value.id;
  }

  get type(): ViewType {
    return this._value.type;
  }

  get title(): string {
    return this._value.title;
  }

  get summary(): string {
    return this._value.summary;
  }

  get description(): string {
    return this._value.description;
  }

  get search(): Search {
    return this._value.search;
  }

  get properties(): Properties {
    return this._value.properties;
  }

  get state(): ViewStateMap {
    return this._value.state;
  }

  get createdAt(): Date {
    return this._value.createdAt;
  }

  get widgetMapping(): WidgetMapping {
    return (this.state || Immutable.Map()).valueSeq().map((s) => s.widgetMapping).reduce((prev, cur) => Immutable.fromJS(prev).merge(Immutable.fromJS(cur)));
  }

  get owner(): string {
    return this._value.owner;
  }

  get requires(): Requirements {
    return this._value.requires || {};
  }

  getSearchTypeByWidgetId(widgetId: string): QuerySearchType | undefined | null {
    const widgetMapping = this.state.map((state) => state.widgetMapping).flatten(true);
    const searchTypeId = widgetMapping.get(widgetId).first();

    if (!searchTypeId) {
      throw new Error(`Search type for widget with id ${widgetId} does not exist`);
    }

    const searchTypes = flatten(this.search.queries.map((query) => query.searchTypes).toArray());

    return searchTypes.find((entry) => entry && entry.id && entry.id === searchTypeId);
  }

  getWidgetTitleByWidget(widget: Widget) {
    const widgetTitles = this.state.flatMap((state) => state.titles.get('widget'));
    const _defaultTitle = defaultTitle(widget);

    return widgetTitles.get(widget.id) || _defaultTitle;
  }

  toBuilder(): Builder {
    const { id, title, summary, description, search, properties, state, createdAt, owner, requires, type } = this._value;

    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder(Immutable.Map({
      id,
      title,
      summary,
      description,
      search,
      properties,
      state,
      createdAt,
      owner,
      requires,
      type,
    }));
  }

  toJSON() {
    const { id, type, title, summary, description, search, properties, state, createdAt, owner } = this._value;

    return {
      id,
      type,
      title,
      summary,
      description,
      search_id: search.id,
      properties,
      state,
      created_at: createdAt,
      owner,
    } as unknown as ViewJson;
  }

  static fromJSON(value: ViewJson): View {
    const { id, type, title, summary, description, properties, state, created_at, owner, requires } = value;
    const viewState: ViewStateMap = Immutable.Map(state).map(ViewState.fromJSON).toMap();

    return View.create()
      .toBuilder()
      .id(id)
      .type(type)
      .title(title)
      .summary(summary)
      .description(description)
      .properties(properties)
      .state(viewState)
      .createdAt(created_at)
      .owner(owner)
      .requires(requires)
      .build();
  }

  static builder(): Builder {
    // eslint-disable-next-line @typescript-eslint/no-use-before-define
    return new Builder();
  }
}

type InternalBuilderState = Immutable.Map<string, any>;

class Builder {
  value: InternalBuilderState;

  constructor(value: InternalBuilderState = Immutable.Map()) {
    this.value = value;
  }

  id(value: string): Builder {
    return new Builder(this.value.set('id', value));
  }

  toNewView(): Builder {
    return new Builder(this.value.set('id', undefined).set('title', undefined));
  }

  type(value: ViewType): Builder {
    return new Builder(this.value.set('type', value));
  }

  newId(): Builder {
    return this.id(new ObjectID().toString());
  }

  title(value: string): Builder {
    return new Builder(this.value.set('title', value));
  }

  summary(value: string): Builder {
    return new Builder(this.value.set('summary', value));
  }

  description(value: string): Builder {
    return new Builder(this.value.set('description', value));
  }

  search(value: Search): Builder {
    return new Builder(this.value.set('search', value));
  }

  properties(value: Properties): Builder {
    return new Builder(this.value.set('properties', value));
  }

  state(value: (ViewStateMap | { [key: string]: ViewState })): Builder {
    return new Builder(this.value.set('state', Immutable.Map(value)));
  }

  createdAt(value: Date): Builder {
    return new Builder(this.value.set('createdAt', value));
  }

  owner(value: string): Builder {
    return new Builder(this.value.set('owner', value));
  }

  requires(value: Requirements): Builder {
    return new Builder(this.value.set('requires', value));
  }

  build(): View {
    const { id, type, title, summary, description, search, properties, state, createdAt, owner, requires } = this.value.toObject();

    return new View(id, type, title, summary, description, search, properties, state, createdAt, owner, requires);
  }
}
