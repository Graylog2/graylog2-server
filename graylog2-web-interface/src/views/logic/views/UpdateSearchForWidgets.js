// @flow strict
import { get } from 'lodash';

import View from 'views/logic/views/View';
import SearchTypesGenerator from 'views/logic/searchtypes/SearchTypesGenerator';

const UpdateSearchForWidgets = (view: View): View => {
  const { state: states } = view;
  const searchTypes = states.map((state) => SearchTypesGenerator(state.widgets));

  const search = get(view, 'search');
  const newQueries = search.queries
    .map((q) => q.toBuilder().searchTypes(searchTypes.get(q.id, {}).searchTypes).build());
  const newSearch = search.toBuilder().newId().queries(newQueries).build();
  let newView = view.toBuilder().search(newSearch).build();

  searchTypes.map(({ widgetMapping }) => widgetMapping)
    .forEach((widgetMapping, queryId) => {
      const newStates = newView.state;

      if (states.has(queryId)) {
        newView = newView.toBuilder()
          .state(newStates.update(queryId, (state) => state.toBuilder().widgetMapping(widgetMapping).build()))
          .build();
      }
    });

  return newView;
};

export default UpdateSearchForWidgets;
