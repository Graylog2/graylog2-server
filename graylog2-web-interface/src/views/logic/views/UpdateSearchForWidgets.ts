/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import { get } from 'lodash';

import View from 'views/logic/views/View';
import SearchTypesGenerator from 'views/logic/searchtypes/SearchTypesGenerator';

const UpdateSearchForWidgets = (view: View): View => {
  const { state: states } = view;
  const searchTypes = states.map((state) => SearchTypesGenerator(state.widgets));

  const search = get(view, 'search');
  const newQueries = search.queries
    .map((q) => q.toBuilder().searchTypes(searchTypes.get(q.id)?.searchTypes ?? []).build());
  const newSearch = search.toBuilder().newId().queries(newQueries.toOrderedSet()).build();
  let newView = view.toBuilder().search(newSearch).build();

  searchTypes.map(({ widgetMapping }) => widgetMapping)
    .forEach((widgetMapping, queryId) => {
      const newStates = newView.state;

      if (states.has(queryId)) {
        newView = newView.toBuilder()
          .state(newStates.update(queryId, (state) => state.toBuilder().widgetMapping(widgetMapping).build()))
          .build();
      }
    });

  return newView;
};

export default UpdateSearchForWidgets;
