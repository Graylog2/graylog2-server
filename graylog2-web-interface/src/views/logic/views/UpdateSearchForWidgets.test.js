import { readFileSync } from 'fs';
import { dirname } from 'path';

import Search from 'views/logic/search/Search';
import View from 'views/logic/views/View';

import UpdateSearchForWidgets from './UpdateSearchForWidgets';
import Parameter from '../parameters/Parameter';
import ValueParameter from '../parameters/ValueParameter';

const cwd = dirname(__filename);
const readFixture = filename => JSON.parse(readFileSync(`${cwd}/${filename}`).toString());

jest.mock('uuid/v4', () => jest.fn(() => 'dead-beef'));

jest.mock('../Widgets', () => ({
  widgetDefinition: () => ({ searchTypes: () => [{ type: 'pivot' }] }),
}));

jest.mock('../SearchType', () => jest.fn(() => ({
  type: 'pivot',
  handler: jest.fn(),
  defaults: {},
})));

describe('UpdateSearchForWidgets', () => {
  beforeEach(() => {
    Parameter.registerSubtype(ValueParameter.type, ValueParameter);
  });

  it('should generate a new search for the view', () => {
    const viewFixture = View.fromJSON(readFixture('./UpdateSearchForWidgets.View.fixture.json'));
    const searchFixture = Search.fromJSON(readFixture('./UpdateSearchForWidgets.Search.fixture.json'));
    const searchView = viewFixture.toBuilder()
      .search(searchFixture)
      .build();

    const newView = UpdateSearchForWidgets(searchView);
    expect(newView).toMatchSnapshot();
  });
});
