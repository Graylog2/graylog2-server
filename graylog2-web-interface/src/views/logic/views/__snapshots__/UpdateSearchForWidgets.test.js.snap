// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateSearchForWidgets should generate a new search for the view 1`] = `
Object {
  "created_at": "2019-07-05T07:35:42.335Z",
  "description": "",
  "id": "5d6d03909889ecf444527f98",
  "owner": "admin",
  "properties": Immutable.List [],
  "search_id": "new-search-id",
  "state": Immutable.Map {
    "f678e6c5-43b1-4200-b5c3-d33eae164dea": Object {
      "formatting": Object {
        "highlighting": Array [],
      },
      "positions": Immutable.Map {
        "98ccf13e-6dec-4362-bcf3-a4776f457061": Object {
          "col": 1,
          "height": 2,
          "row": 7,
          "width": "Infinity",
        },
        "ef9e676a-2d08-4433-b082-2617878a6e59": Object {
          "col": 1,
          "height": 6,
          "row": 9,
          "width": "Infinity",
        },
        "c78e9a51-2e36-4ec4-86de-9df4d29f0863": Object {
          "col": 1,
          "height": 2,
          "row": 1,
          "width": 4,
        },
        "3d568826-fe3d-413a-99a6-09e51d82c851": Object {
          "col": 9,
          "height": 2,
          "row": 1,
          "width": 4,
        },
        "fb496a41-4358-4bc8-b19f-244c71fe9143": Object {
          "col": 5,
          "height": 2,
          "row": 1,
          "width": 4,
        },
        "897a0961-a548-4651-9ca6-f3d6f21b8379": Object {
          "col": 1,
          "height": 4,
          "row": 3,
          "width": 4,
        },
        "ec997825-6065-4fb7-ba86-427e74109ee4": Object {
          "col": 5,
          "height": 4,
          "row": 3,
          "width": 4,
        },
        "dc4a61b2-26e6-468e-bac2-f7702846ca92": Object {
          "col": 9,
          "height": 4,
          "row": 3,
          "width": 4,
        },
      },
      "selected_fields": Array [
        "source",
        "message",
      ],
      "titles": Immutable.Map {
        "widget": Immutable.Map {
          "98ccf13e-6dec-4362-bcf3-a4776f457061": "Message Count",
          "ef9e676a-2d08-4433-b082-2617878a6e59": "All Messages",
          "c78e9a51-2e36-4ec4-86de-9df4d29f0863": "First Commit",
          "3d568826-fe3d-413a-99a6-09e51d82c851": "Last Commit",
          "fb496a41-4358-4bc8-b19f-244c71fe9143": "Commit Count",
          "897a0961-a548-4651-9ca6-f3d6f21b8379": "Commit per Weekday",
          "ec997825-6065-4fb7-ba86-427e74109ee4": "Work Time",
          "dc4a61b2-26e6-468e-bac2-f7702846ca92": "Project Distribution",
        },
        "tab": Immutable.Map {
          "title": "AuthorOverview",
        },
      },
      "widget_mapping": Immutable.Map {
        "dc4a61b2-26e6-468e-bac2-f7702846ca92": Immutable.Set [
          "dead-beef",
        ],
        "fb496a41-4358-4bc8-b19f-244c71fe9143": Immutable.Set [
          "dead-beef",
        ],
        "ec997825-6065-4fb7-ba86-427e74109ee4": Immutable.Set [
          "dead-beef",
        ],
        "c78e9a51-2e36-4ec4-86de-9df4d29f0863": Immutable.Set [
          "dead-beef",
        ],
        "897a0961-a548-4651-9ca6-f3d6f21b8379": Immutable.Set [
          "dead-beef",
        ],
        "98ccf13e-6dec-4362-bcf3-a4776f457061": Immutable.Set [
          "dead-beef",
        ],
        "3d568826-fe3d-413a-99a6-09e51d82c851": Immutable.Set [
          "dead-beef",
        ],
        "ef9e676a-2d08-4433-b082-2617878a6e59": Immutable.Set [
          "dead-beef",
        ],
      },
      "widgets": Immutable.List [
        Object {
          "config": Object {
            "column_pivots": Array [],
            "formatting_settings": Object {
              "chart_colors": Object {
                "count()": "#455a64",
              },
            },
            "rollup": true,
            "row_pivots": Array [
              Object {
                "config": Object {
                  "limit": 15,
                },
                "field": "project",
                "type": "values",
              },
            ],
            "series": Array [
              Object {
                "config": Object {
                  "name": null,
                },
                "function": "count()",
              },
            ],
            "sort": Array [
              Object {
                "direction": "Descending",
                "field": "count()",
                "type": "series",
              },
            ],
            "visualization": "bar",
            "visualization_config": null,
          },
          "filter": undefined,
          "id": "dc4a61b2-26e6-468e-bac2-f7702846ca92",
          "query": null,
          "streams": Array [],
          "timerange": null,
          "type": "aggregation",
        },
        Object {
          "config": Object {
            "column_pivots": Array [],
            "rollup": true,
            "row_pivots": Array [],
            "series": Array [
              Object {
                "config": Object {
                  "name": "Message Count",
                },
                "function": "count()",
              },
            ],
            "sort": Array [],
            "visualization": "numeric",
            "visualization_config": null,
          },
          "filter": undefined,
          "id": "fb496a41-4358-4bc8-b19f-244c71fe9143",
          "query": null,
          "streams": Array [],
          "timerange": null,
          "type": "aggregation",
        },
        Object {
          "config": Object {
            "column_pivots": Array [],
            "formatting_settings": Object {
              "chart_colors": Object {
                "count()": "#fbc02d",
              },
            },
            "rollup": true,
            "row_pivots": Array [
              Object {
                "config": Object {
                  "limit": 24,
                },
                "field": "hour",
                "type": "values",
              },
            ],
            "series": Array [
              Object {
                "config": Object {
                  "name": null,
                },
                "function": "count()",
              },
            ],
            "sort": Array [
              Object {
                "direction": "Ascending",
                "field": "hour",
                "type": "pivot",
              },
            ],
            "visualization": "bar",
            "visualization_config": null,
          },
          "filter": undefined,
          "id": "ec997825-6065-4fb7-ba86-427e74109ee4",
          "query": null,
          "streams": Array [],
          "timerange": null,
          "type": "aggregation",
        },
        Object {
          "config": Object {
            "column_pivots": Array [],
            "rollup": true,
            "row_pivots": Array [],
            "series": Array [
              Object {
                "config": Object {
                  "name": null,
                },
                "function": "min(timestamp)",
              },
            ],
            "sort": Array [],
            "visualization": "numeric",
            "visualization_config": null,
          },
          "filter": undefined,
          "id": "c78e9a51-2e36-4ec4-86de-9df4d29f0863",
          "query": null,
          "streams": Array [],
          "timerange": null,
          "type": "aggregation",
        },
        Object {
          "config": Object {
            "column_pivots": Array [],
            "formatting_settings": Object {
              "chart_colors": Object {
                "count()": "#689f38",
              },
            },
            "rollup": true,
            "row_pivots": Array [
              Object {
                "config": Object {
                  "limit": 15,
                },
                "field": "dayOfTheWeek",
                "type": "values",
              },
              Object {
                "config": Object {
                  "limit": 15,
                },
                "field": "weekday",
                "type": "values",
              },
            ],
            "series": Array [
              Object {
                "config": Object {
                  "name": null,
                },
                "function": "count()",
              },
            ],
            "sort": Array [
              Object {
                "direction": "Ascending",
                "field": "dayOfTheWeek",
                "type": "pivot",
              },
            ],
            "visualization": "bar",
            "visualization_config": null,
          },
          "filter": undefined,
          "id": "897a0961-a548-4651-9ca6-f3d6f21b8379",
          "query": null,
          "streams": Array [],
          "timerange": null,
          "type": "aggregation",
        },
        Object {
          "config": Object {
            "column_pivots": Array [],
            "rollup": true,
            "row_pivots": Array [
              Object {
                "config": Object {
                  "interval": Object {
                    "scaling": null,
                    "type": "auto",
                  },
                },
                "field": "timestamp",
                "type": "time",
              },
            ],
            "series": Array [
              Object {
                "config": Object {
                  "name": null,
                },
                "function": "count()",
              },
            ],
            "sort": Array [],
            "visualization": "bar",
            "visualization_config": null,
          },
          "filter": undefined,
          "id": "98ccf13e-6dec-4362-bcf3-a4776f457061",
          "query": null,
          "streams": Array [],
          "timerange": null,
          "type": "aggregation",
        },
        Object {
          "config": Object {
            "column_pivots": Array [],
            "rollup": true,
            "row_pivots": Array [],
            "series": Array [
              Object {
                "config": Object {
                  "name": null,
                },
                "function": "max(timestamp)",
              },
            ],
            "sort": Array [],
            "visualization": "numeric",
            "visualization_config": null,
          },
          "filter": undefined,
          "id": "3d568826-fe3d-413a-99a6-09e51d82c851",
          "query": null,
          "streams": Array [],
          "timerange": null,
          "type": "aggregation",
        },
        Object {
          "config": Object {
            "fields": Array [
              "timestamp",
              "source",
              "lines_add",
              "lines_removed",
            ],
            "show_message_row": true,
          },
          "filter": undefined,
          "id": "ef9e676a-2d08-4433-b082-2617878a6e59",
          "query": null,
          "streams": Array [],
          "timerange": null,
          "type": "messages",
        },
      ],
    },
  },
  "summary": "",
  "title": "Author Commit History",
  "type": "SEARCH",
}
`;
